File: data structures and algorithms/cpp/08_072.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <algorithm>

using namespace std;

// Structure to represent a restaurant customer
struct Customer {
   string id;
   int partySize;
   int vipStars;
   bool hasReservation;
   int arrivalTime;
   
   // Constructor
   Customer(string id, int partySize, int vipStars, 
         bool hasReservation, int arrivalTime)
      : id(id), partySize(partySize), vipStars(vipStars), 
        hasReservation(hasReservation), 
        arrivalTime(arrivalTime) {}
   
   // Calculate customer priority based on various factors
   int getPriority(int currentTime) const {
      int waitingTime = currentTime - arrivalTime;
      return (vipStars * 10) + 
             (hasReservation ? 20 : 0) + 
             waitingTime;
   }
};

// Structure to represent seating result
struct Result {
   string customerId;
   int seatedTime;
   
   // Constructor
   Result(string customerId, int seatedTime)
      : customerId(customerId), seatedTime(seatedTime) {}
};

// Custom comparator for the priority queue
class CustomerComparator {
private:
   int currentTime;
   
public:
   CustomerComparator(int time) : currentTime(time) {}
   
   // Compare customers by priority
   bool operator()(const Customer& c1, 
                  const Customer& c2) const {
      // In C++ priority_queue, use "less" for max-heap
      return c1.getPriority(currentTime) < 
             c2.getPriority(currentTime);
   }
};

// Process restaurant events and seat customers
vector<Result> restaurantPrioritySeating(
   const vector<int>& tables, 
   const vector<vector<variant<string, int, bool>>>& events) {
   
   // Sort events by time
   vector<vector<variant<string, int, bool>>> sortedEvents = 
      events;
   sort(sortedEvents.begin(), sortedEvents.end(), 
        [](const auto& e1, const auto& e2) { 
           return get<int>(e1[1]) < get<int>(e2[1]); 
        });
   
   vector<Result> results;
   int currentTime = 0;
   priority_queue<Customer, vector<Customer>, 
                 CustomerComparator> waitlist(
                    CustomerComparator(currentTime));
   
   for (const auto& event : sortedEvents) {
      string eventType = get<string>(event[0]);
      currentTime = get<int>(event[1]);
      
      // Update priority queue with new time
      vector<Customer> tempCustomers;
      while (!waitlist.empty()) {
         tempCustomers.push_back(waitlist.top());
         waitlist.pop();
      }
      
      priority_queue<Customer, vector<Customer>, 
                    CustomerComparator> updatedWaitlist(
                       CustomerComparator(currentTime));
      for (const auto& customer : tempCustomers) {
         updatedWaitlist.push(customer);
      }
      waitlist = move(updatedWaitlist);
      
      if (eventType == "ARRIVE") {
         string customerId = get<string>(event[2]);
         int partySize = get<int>(event[3]);
         int vipStars = get<int>(event[4]);
         bool hasReservation = get<bool>(event[5]);
         
         Customer newCustomer(customerId, partySize, 
                            vipStars, hasReservation, 
                            currentTime);
         waitlist.push(newCustomer);
      } 
      else if (eventType == "TABLE_READY") {
         int tableSize = get<int>(event[2]);
         
         vector<Customer> temp;
         bool seated = false;
         
         while (!waitlist.empty() && !seated) {
            Customer customer = waitlist.top();
            waitlist.pop();
            
            if (customer.partySize <= tableSize) {
               results.push_back(
                  Result(customer.id, currentTime));
               seated = true;
            } else {
               temp.push_back(customer);
            }
         }
         
         // Restore customers we examined but didn't seat
         for (const auto& customer : temp) {
            waitlist.push(customer);
         }
      }
   }
   
   return results;
}

File: data structures and algorithms/cs/08_072.cs
using System;
using System.Collections.Generic;

class Customer
{
   public string Id { get; }
   public int PartySize { get; }
   public int VipStars { get; }
   public bool HasReservation { get; }
   public int ArrivalTime { get; }

   public Customer(
      string id, 
      int partySize, 
      int vipStars,
      bool hasReservation, 
      int arrivalTime)
   {
      Id = id;
      PartySize = partySize;
      VipStars = vipStars;
      HasReservation = hasReservation;
      ArrivalTime = arrivalTime;
   }

   public int GetPriority(int currentTime)
   {
      int waitingTime = currentTime - ArrivalTime;
      return (VipStars * 10) + 
         (HasReservation ? 20 : 0) +
         waitingTime;
   }
}

class Result
{
   public string CustomerId { get; }
   public int SeatedTime { get; }

   public Result(string customerId, int seatedTime)
   {
      CustomerId = customerId;
      SeatedTime = seatedTime;
   }
}

public class RestaurantPrioritySeating
{
   public static List<Result> ManageSeating(
      List<int> tables,
      List<object[]> events)
   {
      events.Sort((e1, e2) => 
         ((int)e1[1]).CompareTo((int)e2[1]));

      var waitlist = new PriorityQueue<Customer>(
         Comparer<Customer>.Create((c1, c2) =>
            c2.GetPriority(0).CompareTo(c1.GetPriority(0))));

      List<Result> results = new List<Result>();

      foreach (object[] eventObj in events)
      {
         string eventType = (string)eventObj[0];
         int currentTime = (int)eventObj[1];

         var updatedWaitlist = new PriorityQueue<Customer>(
            Comparer<Customer>.Create((c1, c2) => 
               c2.GetPriority(currentTime)
                  .CompareTo(c1.GetPriority(currentTime))));

         foreach (Customer customer in waitlist)
         {
            updatedWaitlist.Enqueue(customer);
         }
         waitlist = updatedWaitlist;

         if (eventType == "ARRIVE")
         {
            string customerId = (string)eventObj[2];
            int partySize = (int)eventObj[3];
            int vipStars = (int)eventObj[4];
            bool hasReservation = (bool)eventObj[5];

            Customer newCustomer = new Customer(
               customerId, 
               partySize, 
               vipStars,
               hasReservation, 
               currentTime);
            waitlist.Enqueue(newCustomer);
         }
         else if (eventType == "TABLE_READY")
         {
            int tableSize = (int)eventObj[2];
            List<Customer> temp = new List<Customer>();
            Customer seated = null;

            while (waitlist.Count > 0 && seated == null)
            {
               Customer customer = waitlist.Dequeue();

               if (customer.PartySize <= tableSize)
               {
                  seated = customer;
                  results.Add(
                     new Result(
                        customer.Id,
                        currentTime));
               }
               else
               {
                  temp.Add(customer);
               }
            }

            foreach (Customer customer in temp)
            {
               waitlist.Enqueue(customer);
            }
         }
      }

      return results;
   }
}
